### PR TITLE
Replace Ledger.Tx.ChainIndexTxOut

### PR DESCRIPTION
**Pull request by andreabedini**
tl;dr. 

```
data DecoratedTxOut = DecoratedTxOut
  { decoratedTxOutAddress         :: Address
  , decoratedTxOutValue           :: Value
  , decoratedTxOutValidator       :: Maybe (Either ValidatorHash Validator)
  , decoratedTxOutDatum           :: Maybe (Either DatumHash Datum)
  , decoratedTxOutReferenceScript :: Maybe Script
  }
```

The idea is that at the contract level, we need a `TxOut` with additional information like the resolved validator & datum. I stole the word "decorated" from @michaelpj but I am happy to consider alternatives. In principle one could stick with `TxOut` and rely on qualified names.

Taking a  KISS approach I have defined a type that has all the information. The type is not "tight" sinceit has invariants:

If the address corresponds to a script then
1. the validator field must match
2. the datum must be present

(Note that `decoratedTxOutDatum` still returns a `Maybe`, perhaps we can improve on this).

Therefore the `DecoratedTxOut` constructor is private. I considered other internal representations but I am tempted to stick with this

There are few smart constructors exported:

```
mkDecoratedTxOut :: Address -> Value -> Maybe (Either DatumHash Datum) -> Maybe Script -> DecoratedTxOut
mkDecoratedTxOutFromValidator :: Validator -> Value -> Either DatumHash Datum -> Maybe Script -> DecoratedTxOut
mkDecoratedTxOutFromValidatorHash :: ValidatorHash -> Value -> Either DatumHash Datum -> Maybe Script -> DecoratedTxOut
```

Also there's a function to turn a `DecoratedTxOut` into a `TxOut`. I don't like this function because the type offers no guarantee to be equivalent to a `TxOut` but for the moment we need it because `plutus-script-utils` is lower level than `plutus-ledger` and takes `TxOut`s.

```
decoratedTxOutToV1 :: DecoratedTxOut -> TxOut
```

`Plutus.ChainIndex.Types.ChainIndexTxOut` is left alone as a internal type for plutus-chain-index. The chain-index API talks to the exernal parties thorugh `DecoratedTxOut` as defined in `plutus-ledger`.

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reference the ADR in the PR and reference the PR in the ADR (if revelant)
    - [ ] Reviewer requested

### CODE REVIEW

**Comment by andreabedini** [comment]
After an internal discussion I have changed the design to

```
data OffChainTxOut =
    PublicKeyOffChainTxOut' {
      -- | Hash of the public key protecting the transaction output.
      ocTxOutAddress            :: V1.PubKeyHash,
      -- | Value of the transaction output.
      ocTxOutValue              :: V1.Value,
      -- | Optional datum hash attached to the transaction output.
      ocTxOutPublicKeyDatumHash :: Maybe V1.DatumHash,
      -- | Optional datum attached to the transaction output, either found
      -- inline or resolved from its hash offchain.
      ocTxOutDatum              :: Maybe V1.Datum,
      -- | Optional reference script attached to the transaction output.
      ocTxOutReferenceScript    :: Maybe V1.Script
    }
  | ScriptOffChainTxOut' {
      -- | Hash of the validator protecting the transaction output.
      ocTxOutValidatorHash   :: V1.ValidatorHash,
      -- | Value of the transaction output.
      ocTxOutValue           :: V1.Value,
      -- | FIXME Hash of the datum attached to the transaction output.
      ocTxOutScriptDatumHash :: V1.DatumHash,
      -- | Validator script protecting the transaction output, resolved
      -- from its hash off-chain
      ocTxOutValidator       :: Maybe V1.Validator,
      -- | Datum attached to the transaction output, either found inline or
      -- resolved from its hash off-chain. A transaction output protected
      -- by a Plutus script is need to have an associated datum to be
      -- spendable.
      ocTxOutDatum           :: Maybe V1.Datum,
      -- | Optional reference script attached to the transaction output.
      -- The reference script is, in genereal, unrelated to the validator
      -- script althought it could also be the same.
      ocTxOutReferenceScript :: Maybe V1.Script
    }
  deriving (Show, Eq, Serialise, Generic, ToJSON, FromJSON, OpenApi.ToSchema)
```

`PublicKeyOffChainTxOut'` and `ScriptOffChainTxOut'` are hidden and only unidirectional patterns are exported (`PublicKeyOffChainTxOut`, `ScriptOffChainTxOut`).

I got some smart constructors to make sure validator and datum match their hash.

```
mkPublicKeyOffChainTxOut :: V1.PubKeyHash -> V1.Value -> Maybe (Either V1.DatumHash V1.Datum) -> Maybe V1.Script -> OffChainTxOut
mkScriptOffChainTxOut :: Either V1.ValidatorHash V1.Validator -> V1.Value -> Either V1.DatumHash V1.Datum -> Maybe V1.Script -> OffChainTxOut
```

I also wrote a small accessor for datums, which seemed handy
```
ocTxOutDatumHash :: OffChainTxOut -> Maybe V1.DatumHash
ocTxOutDatumHash (PublicKeyOffChainTxOut' _pkh _va m_dh _m_da _m_sc)  = m_dh
ocTxOutDatumHash (ScriptOffChainTxOut' _vh _va dh _m_val _m_da _m_sc) = Just dh
```

File: plutus-contract/src/Plutus/Contract/StateMachine.hs [hunk 1]
```diff
@@ -135,13 +135,13 @@ getStates
...
     -> [OnChainState s i]
 getStates (SM.StateMachineInstance _ si) refMap =
-    flip mapMaybe (Map.toList refMap) $ \(txOutRef, ciTxOut) -> do
-      let txOut = Tx.toTxOut ciTxOut
-      datum <- ciTxOut ^? Tx.ciTxOutScriptDatum . _Right
-      ocsTxOutRef <- either (const Nothing) Just $ Typed.typeScriptTxOutRef si txOutRef txOut datum
+    flip mapMaybe (Map.toList refMap) $ \(txOutRef, ocTxOut) -> do
+      -- FIXME is it important to check txOut is a script?
+      datum <- Tx.ocTxOutDatum ocTxOut
+      ocsTxOutRef <- either (const Nothing) Just $ Typed.typeScriptTxOutRef si txOutRef (Tx.toTxOut ocTxOut) datum
```
Comment (andreabedini) [inline plutus-contract/src/Plutus/Contract/StateMachine.hs:144]: This funtion used `ciTxOutScriptDatum` which would only match when the txOut is protected by a script. Is this important? `ocTxOutDatum` on the other hand gives you a `Maybe Datum` independently from whether the txOut is protected by a pubkey or a script.

File: plutus-playground-server/usecases/Game.hs [hunk 2]
```diff
@@ -137,14 +137,15 @@ guess = endpoint @"guess" @GuessParams $ \(GuessParams theGuess) -> do
...
 findSecretWordValue =
   listToMaybe . catMaybes . Map.elems . Map.map secretWordValue
 
 -- | Extract the secret word in the Datum of a given transaction output is possible
-secretWordValue :: ChainIndexTxOut -> Maybe HashedString
+secretWordValue :: OffChainTxOut -> Maybe HashedString
 secretWordValue o = do
-  Datum d <- either (const Nothing) Just (_ciTxOutScriptDatum o)
+  -- FIXME is it important to check txOut is a script?
+  Datum d <- ocTxOutDatum o
```
Comment (andreabedini) [inline plutus-playground-server/usecases/Game.hs:148]: This funtion used `ciTxOutScriptDatum` which would only match when the txOut is protected by a script. Is this important? `ocTxOutDatum` on the other hand gives you a `Maybe Datum` independently from whether the txOut is protected by a pubkey or a script.

File: plutus-use-cases/src/Plutus/Contracts/Tutorial/Escrow.hs [hunk 3]
```diff
@@ -302,7 +302,11 @@ refund ::
 refund inst _escrow = do
     pk <- ownFirstPaymentPubKeyHash
     unspentOutputs <- utxosAt (Scripts.validatorAddress inst)
-    let flt _ ciTxOut = either id Ledger.datumHash (Tx._ciTxOutScriptDatum ciTxOut) == Ledger.datumHash (Datum (PlutusTx.toBuiltinData pk))
+    let flt _ ocTxOut = case ocTxOut of
+                          Tx.ScriptOffChainTxOut _vh _v dh _m_va _m_da _m_rs ->
+                            dh == Ledger.datumHash (Datum (PlutusTx.toBuiltinData pk))
+                          Tx.PublicKeyOffChainTxOut {} ->
+                            False
```
Comment (andreabedini) [inline plutus-use-cases/src/Plutus/Contracts/Tutorial/Escrow.hs:309]: Same here, it's not clear whether it's relevent that the txOut is protected by a script or pubkey.

File: plutus-contract/src/Plutus/Contract/StateMachine.hs (hunk 1 above)
Comment (sjoerdvisscher) [reply plutus-contract/src/Plutus/Contract/StateMachine.hs:144]: I checked, I'm pretty sure this function only gets to see script outputs.

File: plutus-playground-server/usecases/Game.hs (hunk 2 above)
Comment (sjoerdvisscher) [reply plutus-playground-server/usecases/Game.hs:148]: Same here.

File: plutus-use-cases/src/Plutus/Contracts/Tutorial/Escrow.hs (hunk 3 above)
Comment (sjoerdvisscher) [reply plutus-use-cases/src/Plutus/Contracts/Tutorial/Escrow.hs:309]: Here it's the easiest to check, `flt` is applied to the result of `unspentOutputs <- utxosAt (Scripts.validatorAddress inst)`.

File: plutus-contract/src/Plutus/Contract/StateMachine.hs (hunk 1 above)
Comment (andreabedini) [reply plutus-contract/src/Plutus/Contract/StateMachine.hs:144]: Ok, so we can keep the `ocTxOutDatum` since at this point the difference is inessential right?

File: plutus-use-cases/src/Plutus/Contracts/Tutorial/Escrow.hs (hunk 3 above)
Comment (andreabedini) [reply plutus-use-cases/src/Plutus/Contracts/Tutorial/Escrow.hs:309]: ok so I can change this to `\ocTxOut -> ocTxOutDatum ocTxOut == Just (Ledger.datumHash (Datum (PlutusTx.toBuiltinData pk)))` ?

File: plutus-use-cases/src/Plutus/Contracts/Escrow.hs [hunk 4]
```diff
@@ -332,7 +332,11 @@ refund ::
 refund inst escrow = do
     pk <- ownFirstPaymentPubKeyHash
     unspentOutputs <- utxosAt (Scripts.validatorAddress inst)
-    let flt _ ciTxOut = either id datumHash (Tx._ciTxOutScriptDatum ciTxOut) == datumHash (Datum (PlutusTx.toBuiltinData pk))
+    let flt _ ocTxOut = case ocTxOut of
+                          Tx.ScriptOffChainTxOut _vh _v dh _m_va _m_da _m_rs ->
+                            dh == datumHash (Datum (PlutusTx.toBuiltinData pk))
```
[the remark this answers is not in the document]
Comment (sjoerdvisscher) [reply plutus-use-cases/src/Plutus/Contracts/Escrow.hs:337]: The same in many other places of course.

File: plutus-chain-index-core/src/Plutus/ChainIndex/Emulator/Handlers.hs [hunk 5]
```diff
@@ -123,33 +105,51 @@ getUtxoutFromRef ::
...
+  -> Eff effs (Maybe L.OffChainTxOut)
+makeOffChainTxOut txOut@(ChainIndexTxOut address value datum refScript) = do
+  ds <- gets (view diskState)
+  -- The output might come from a public key address or a script address.
+  -- We need to handle them differently.
+  case addressCredential address of
+    PubKeyCredential pkh ->
+      case datum of
+        OutputDatumHash dh ->
+          pure $ Just $ L.mkPublicKeyOffChainTxOut pkh value (Just (Left dh)) (fromReferenceScript refScript)
```
Comment (sjoerdvisscher) [inline plutus-chain-index-core/src/Plutus/ChainIndex/Emulator/Handlers.hs:134]: Why not change `mkPublicKeyOffChainTxOut` to use the `OutputDatum` type?

**Review by sjoerdvisscher — CHANGES_REQUESTED** [review]

File: plutus-use-cases/src/Plutus/Contracts/Tutorial/Escrow.hs (hunk 3 above)
Comment (sjoerdvisscher) [reply plutus-use-cases/src/Plutus/Contracts/Tutorial/Escrow.hs:309]: Yes

File: plutus-contract/src/Plutus/Contract/StateMachine.hs (hunk 1 above)
Comment (sjoerdvisscher) [reply plutus-contract/src/Plutus/Contract/StateMachine.hs:144]: Yes

File: plutus-ledger/src/Ledger/Tx.hs [hunk 6]
```diff
@@ -75,99 +73,119 @@ import Data.Proxy (Proxy (Proxy))
...
-    PublicKeyChainIndexTxOut {
-      -- | Address of the transaction output. The address is protected by a
-      -- public key hash.
-      _ciTxOutAddress         :: Address,
+-- | Transaction output represented off-chain
+-- add more
+data OffChainTxOut =
+    PublicKeyOffChainTxOut' {
+      -- | Hash of the public key protecting the transaction output.
+      ocTxOutAddress            :: V1.PubKeyHash,
```
Comment (koslambrou) [inline plutus-ledger/src/Ledger/Tx.hs:98]: `V1.Address` instead of `V1.PubKeyHash`

File: plutus-ledger/src/Ledger/Tx.hs [hunk 7]
```diff
@@ -75,99 +73,119 @@ import Data.Proxy (Proxy (Proxy))
...
       -- | Value of the transaction output.
-      _ciTxOutValue           :: V1.Value,
-      -- | Optional datum attached to the transaction output.
-      _ciTxOutPublicKeyDatum  :: V2.OutputDatum,
+      ocTxOutValue              :: V1.Value,
+      -- | Optional datum hash attached to the transaction output.
+      ocTxOutPublicKeyDatumHash :: Maybe V1.DatumHash,
+      -- | Optional datum attached to the transaction output, either found
+      -- inline or resolved from its hash offchain.
+      ocTxOutDatum              :: Maybe V1.Datum,
```
Comment (koslambrou) [inline plutus-ledger/src/Ledger/Tx.hs:105]: Why into 2 separate fields instead of a Tuple?

File: plutus-ledger/src/Ledger/Tx.hs [hunk 8]
```diff
@@ -75,99 +73,119 @@ import Data.Proxy (Proxy (Proxy))
...
-      _ciTxOutReferenceScript :: Maybe V1.Script,
-      -- | Validator protecting the transaction output, either in full or
-      -- as a hash reference.
-      _ciTxOutValidator       :: Either V1.ValidatorHash V1.Validator
+      ocTxOutReferenceScript :: Maybe V1.Script
     }
   deriving (Show, Eq, Serialise, Generic, ToJSON, FromJSON, OpenApi.ToSchema)
 
-makeLenses ''ChainIndexTxOut
-makePrisms ''ChainIndexTxOut
```
Comment (koslambrou) [inline plutus-ledger/src/Ledger/Tx.hs:137]: Should we bring those? We use `lens` in our codebase, so having them seems useful.

File: plutus-ledger/src/Ledger/Tx.hs [hunk 9]
```diff
@@ -75,99 +73,119 @@ import Data.Proxy (Proxy (Proxy))
...
-      _ciTxOutReferenceScript :: Maybe V1.Script
+      ocTxOutReferenceScript    :: Maybe V1.Script
     }
-  | ScriptChainIndexTxOut {
-      -- | Address of the transaction output. The address is protected by a
-      -- script.
-      _ciTxOutAddress         :: Address,
+  | ScriptOffChainTxOut' {
+      -- | Hash of the validator protecting the transaction output.
+      ocTxOutValidatorHash   :: V1.ValidatorHash,
```
Comment (koslambrou) [inline plutus-ledger/src/Ledger/Tx.hs:111]: `Address` and not `V1.ValidatorHash`.

File: plutus-ledger/src/Ledger/Tx.hs [hunk 10]
```diff
@@ -75,99 +73,119 @@ import Data.Proxy (Proxy (Proxy))
...
--- 'Ledger.Constraints.processConstraint', specifically with the constraints
--- 'MustSpendPubKeyOutput' and 'MustSpendScriptOutput'.
-data ChainIndexTxOut =
-    PublicKeyChainIndexTxOut {
-      -- | Address of the transaction output. The address is protected by a
-      -- public key hash.
-      _ciTxOutAddress         :: Address,
+-- | Transaction output represented off-chain
+-- add more
+data OffChainTxOut =
```
Comment (koslambrou) [inline plutus-ledger/src/Ledger/Tx.hs:95]: Curious as to why you went from `DecoratedTxOut` to `OffChainTxOut`?

File: plutus-ledger/src/Ledger/Tx.hs (hunk 7 above)
Comment (andreabedini) [reply plutus-ledger/src/Ledger/Tx.hs:105]: Because you might have a `DatumHash` but not a `Datum`. It would be tighter to use `Maybe (DatumHash, Maybe Datum)` since you cannot have a Datum without its hash.

File: doc/plutus/tutorials/EscrowImpl.hs [hunk 11]
```diff
@@ -308,7 +308,7 @@ redeem inst escrow = mapError (review _EscrowError) $ do
                 <> Constraints.mustValidateIn valRange
     if current >= escrowDeadline escrow
     then throwing _RedeemFailed DeadlinePassed
-    else if foldMap (view Tx.ciTxOutValue) unspentOutputs `lt` targetTotal escrow
```
Comment (raduom) [inline doc/plutus/tutorials/EscrowImpl.hs:311]: Why are we dropping lenses?

File: plutus-chain-index-core/src/Plutus/ChainIndex/Emulator/Handlers.hs (hunk 5 above)
Comment (andreabedini) [reply plutus-chain-index-core/src/Plutus/ChainIndex/Emulator/Handlers.hs:134]: Good question, `OutputDatum` is only present in the V2 `plutus-ledger-api` and I wanted to avoid having to expose V2 stuff to all code consuming `plutus-ledger` (in this particular case it's chain-index that decided to use `OutputDatum`, other representations were possible).

File: plutus-ledger/src/Ledger/Tx.hs (hunk 6 above)
Comment (andreabedini) [reply plutus-ledger/src/Ledger/Tx.hs:98]: I actually meant to use `PubKeyHash` here, it's the name that's wrong. `ocTxOutAddress` will be an accessor function `ocTxOutAddress :: OffChainTxOut -> Address`.

Comment (andreabedini) [reply plutus-ledger/src/Ledger/Tx.hs:98]: This is because `PubKeyHash` is actual evidence the txOut is protected by a public key, while `Address` can stil be a script.

File: plutus-ledger/src/Ledger/Tx.hs (hunk 10 above)
Comment (andreabedini) [reply plutus-ledger/src/Ledger/Tx.hs:95]: I got some feedbeck that
1) having two constructors is useful
2) given this type represents a txOut as seen off-chain `OffChainTxOut` is a good name. `TxOut` is the on-chain version.
3) someone suggested "isn't just renaming enough", and this version is a simpler approch, mostly a rename with some tightening of the types.

Basically `OffChainTxOut` is made by
1) embedding the two constructors of `Credential` into the type, to retain evidence of the type of the address.
2) Turning `Either V1.ValidatorHash V1.Validator` -> `Maybe V1.Validator` since we already have the validator hash thanks to the above
3) Split `Either V1.DatumHash V1.Datum` into `Maybe DatumHash`, `Maybe Datum` to keep them both so we don't have to recompute the hash.
4) Expand `V2.OutputDatum` so we don't have to depend on V2, since we actually don't need to.

File: plutus-ledger/src/Ledger/Tx.hs (hunk 9 above)
Comment (andreabedini) [reply plutus-ledger/src/Ledger/Tx.hs:111]: Same as [above](https://github.com/input-output-hk/plutus-apps/pull/626#discussion_r931918073)

File: plutus-ledger-constraints/src/Ledger/Constraints/OffChain.hs [hunk 12]
```diff
@@ -710,15 +711,18 @@ resolveScriptTxOut
...
+resolveScriptTxOut (Tx.PublicKeyOffChainTxOut _pkh _va _m_dh _m_da _m_sc) =
+  pure Nothing
+resolveScriptTxOut (Tx.ScriptOffChainTxOut vh va dh m_val m_da _m_sc) = do
+  -- If we don't have a validator already, look for it in the 'slOtherScripts map.
+  validator <- case m_val of
+    Just val -> pure val
+    Nothing  -> lookupValidator vh
+
+  -- Same for the datum
+  dataValue <- case m_da of
```
Comment (raduom) [inline plutus-ledger-constraints/src/Ledger/Constraints/OffChain.hs:724]: Should this be `datum`?

**Review by raduom — APPROVED** [review]
LGTM

File: plutus-chain-index-core/src/Plutus/ChainIndex/Emulator/Handlers.hs (hunk 5 above)
Comment (sjoerdvisscher) [reply plutus-chain-index-core/src/Plutus/ChainIndex/Emulator/Handlers.hs:134]: Another option would be to have a `OutputDatum -> Maybe (Either DatumHash Datum)` function, to avoid too much repetition here. (The rest seems to be the same, but I might overlook something)

File: plutus-ledger/src/Ledger/Tx.hs (hunk 6 above)
Comment (koslambrou) [reply plutus-ledger/src/Ledger/Tx.hs:98]: The problem is that were losing stake information. I don't see the problem of simply using `Address` here instead of requiring users to reconstruct the `Address` with the PubKeyHash and stake information everytime we want to get the address. Users would need to use `Address` instead of `PubKeyHash`

File: plutus-ledger-constraints/src/Ledger/Constraints/OffChain.hs (hunk 12 above)
Comment (andreabedini) [reply plutus-ledger-constraints/src/Ledger/Constraints/OffChain.hs:724]: Possibly, I had not changed the name.

File: plutus-chain-index-core/src/Plutus/ChainIndex/Emulator/Handlers.hs (hunk 5 above)
Comment (andreabedini) [reply plutus-chain-index-core/src/Plutus/ChainIndex/Emulator/Handlers.hs:134]: @sjoerdvisscher I agree. I feel `plutus-ledger-api` could offer more functions to use `OutputDatum` but nobody prevents us from introducing an internal helper here to reduce duplication. :+1:

File: doc/plutus/tutorials/EscrowImpl.hs (hunk 11 above)
Comment (andreabedini) [reply doc/plutus/tutorials/EscrowImpl.hs:311]: It wasn't intentional, I'll bring this up in chat so we can discuss.

File: plutus-ledger/src/Ledger/Tx.hs (hunk 8 above)
Comment (andreabedini) [reply plutus-ledger/src/Ledger/Tx.hs:137]: Let's discuss in chat.

File: plutus-ledger/src/Ledger/Tx.hs (hunk 6 above)
Comment (andreabedini) [reply plutus-ledger/src/Ledger/Tx.hs]: > The problem is that were losing stake information

That can be added back.

> I don't see the problem of simply using Address here instead of requiring users to reconstruct the Address with the PubKeyHash and stake information everytime we want to get the address.

The goal was to make the following invalid state unrepresentable:

```
PublicKeyChainIndexTxOut (scriptHashAddress "abcd") mempty
```

**Comment by andreabedini** [comment]
I am going to put this on hold to give priority to other work. Meanwhile, feel free to leave any comment you might have.

File: plutus-ledger/src/Ledger/Tx.hs (hunk 6 above)
Comment (koslambrou) [reply plutus-ledger/src/Ledger/Tx.hs]: Ah I see. Then we just need to add staking information as a new field?

Comment (andreabedini) [reply plutus-ledger/src/Ledger/Tx.hs]: Yup, we can do that.

**Comment by koslambrou** [comment]
Closing as superceded by https://github.com/input-output-hk/plutus-apps/pull/799
